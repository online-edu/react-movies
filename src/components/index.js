import HeaderBar from './header';
import Poster from './poster';
import Spinner from './spinner';
import Switch from './switch';
import Modal from './modal';
import Filter from './filter';

export { HeaderBar, Spinner, Poster, Modal, Filter, Switch };
