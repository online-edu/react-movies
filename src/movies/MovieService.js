import { api, key } from '../utils/config';
import request from '../utils/request';

/**
 * Load movies from server
 *
 * @returns {Promise}
 */
const loadMovies = () =>
    request.get(`${api}/movie/popular${key}`).then(({ results: movies }) => Promise.resolve(movies));
/**
 * Load movie details by id from server
 *
 * @returns {Promise}
 */
const loadMovieById = id => request.get(`${api}/movie/${id}${key}`);
/**
 * Filter movies on given criteria.
 *
 * @param {Array} movies - an array of movies
 * @param {Object} filters - filter options
 * @returns {Array}
 */
const filterMovies = (movies, filters) => {
    const props = Object.keys(filters);
    return movies.filter(movie =>
        props.every(prop =>
            prop === 'vote_average'
                ? movie[prop] >= filters[prop]
                : movie[prop] === filters[prop],
        ),
    );
};

export { loadMovies, loadMovieById, filterMovies };
