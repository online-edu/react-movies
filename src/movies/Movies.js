import React, { Component } from 'react';
import Row from 'react-bootstrap/Row';
import Col from 'react-bootstrap/Col';
import { Poster, Spinner, Modal, Filter } from '../components';
import { loadMovies, loadMovieById, filterMovies } from './MovieService';
import MovieDetails from './MovieDetails';

/**
 * Movie component.
 */
class Movies extends Component {
    /**
     * Intilalizes the state and binds all methods.
     *
     * @param {Object} props - access 'props'
     */
    constructor(props) {
        super(props);
        this.state = {
            movies: [],
            source: [],
            movie: {},
            loading: true,
            showModal: false,
        };
        this.onMovieClick = this.onMovieClick.bind(this);
        this.onModalClose = this.onModalClose.bind(this);
        this.onFilterChange = this.onFilterChange.bind(this);
    }

    /**
     * Load movies once component is mounted.
     */
    componentDidMount() {
        loadMovies()
            .then(movies => this.setState({ movies, source: movies, loading: false }))
            .catch(err => this.handleError(err));
    }

    /**
     * Get a movie details
     *
     * @param {number} id - movie id
     */
    onMovieClick(id) {
        loadMovieById(id)
            .then(movie => this.setState({ showModal: true, movie, loading: false }))
            .catch(err => this.handleError(err));
    }

    /**
     * Toggle modal state on close.
     *
     */
    onModalClose() {
        this.setState({ showModal: false });
    }

    /**
     * Toggle modal state on close.
     *
     * @param {Object} filters - list of set filters
     */
    onFilterChange(filters) {
        const { source } = this.state;
        const newMovies = filterMovies(source, filters);
        this.setState({ movies: newMovies });
    }

    /**
     * Handles common errors for n/w requests.
     *
     * @param {string} err - Error message
     */
    handleError(err) {
        console.error(err);
        this.setState({ loading: false });
    }

    /**
     * Render method for component
     */
    render() {
        const { movies, loading, showModal, movie } = this.state;
        return (
            <section>
                <Filter onChange={this.onFilterChange} />
                <Row className="justify-content-center">
                    {(!loading
                        && movies.map(m => (
                            <Col className="my-3 text-center" key={m.id}>
                                <Poster
                                    id={m.id}
                                    name={m.original_title}
                                    url={m.poster_path}
                                    click={this.onMovieClick}
                                />
                            </Col>
                        ))) || <Spinner />}
                </Row>
                {showModal && (
                    <MovieDetails
                        show={showModal}
                        onHide={this.onModalClose}
                        movie={movie}
                    />
                )}
            </section>
        );
    }
}

export default Movies;
