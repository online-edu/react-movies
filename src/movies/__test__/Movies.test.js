import React from 'react';
import { shallow } from 'enzyme';
import Movies from '../Movies';
import { loadMovies, loadMovieById } from '../MovieService';
import { Modal, Spinner } from '../../components';

describe('<Movie />', () => {
    it('renders correctly', () => {
        const component = shallow(<Movies />);
        expect(component).toMatchSnapshot();
    });

    it('renders a <Modal /> component', () => {
        const wrapper = shallow(<Movies />);
        expect(wrapper.find(Modal)).toHaveLength(1);
    });

    it('renders a <Spinner /> component', () => {
        const wrapper = shallow(<Movies />);
        expect(wrapper.find(Spinner)).toHaveLength(1);
    });

    it('calls componentDidMount', () => {
        jest.spyOn(Movies.prototype, 'componentDidMount');
        const wrapper = shallow(<Movies />);
        wrapper.update();
        expect(Movies.prototype.componentDidMount).toBeCalledTimes(1);
    });

    it('expects to change state for "showModal" on modal close', () => {
        const wrapper = shallow(<Movies />);
        wrapper.setState({ showModal: true });
        wrapper.instance().onModalClose();
        expect(wrapper.state().showModal).toBe(false);
    });

    it('expects "source, movies" to be defined', () => {
        const wrapper = shallow(<Movies />);
        wrapper.instance().onFilterChange({ original_language: 'en' });
        const { source } = wrapper.state();
        expect(source).toBeDefined();
    });

    it('componentDidMount should fetch, and put movies in state', () => {
        const wrapper = shallow(<Movies />);
        return loadMovies()
            .then(movies => {
                wrapper.setState({ movies });
                expect(wrapper.state('movies').length).toEqual(20);
            })
            .catch(err => {
                wrapper.setState({ loading: false });
                expect(wrapper.state('loading')).toBe(false);
                expect(err).toBeDefined();
            });
    });

    it('onMovieClick should fetch movie and put it in state', () => {
        const wrapper = shallow(<Movies />);
        wrapper.instance().onMovieClick(480530);
        wrapper.update();
        expect(wrapper.state('movies')).toBeDefined();
        return loadMovieById(480530).then(movie => {
            wrapper.setState({ showModal: true, movie, loading: false });
            expect(wrapper.state('movie')).toHaveProperty(['id']);
            expect(wrapper.state('showModal')).toBe(true);
        });
    });
});
