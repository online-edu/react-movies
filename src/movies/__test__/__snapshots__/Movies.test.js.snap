// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Movie /> renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Movies />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Filter
          onChange={[Function]}
        />,
        <ForwardRef(Bootstrap(Row))
          className="justify-content-center"
        >
          <Spinner />
        </ForwardRef(Bootstrap(Row))>,
        false,
        <CommonModal
          onHide={[Function]}
          show={false}
          title="Oops!"
        >
          <strong>
            
          </strong>
        </CommonModal>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "onChange": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": <Spinner />,
          "className": "justify-content-center",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": Object {
          "$$typeof": Symbol(react.forward_ref),
          "defaultProps": undefined,
          "propTypes": undefined,
          "render": [Function],
        },
      },
      false,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": <strong>
            
          </strong>,
          "onHide": [Function],
          "show": false,
          "title": "Oops!",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "",
          },
          "ref": null,
          "rendered": "",
          "type": "strong",
        },
        "type": [Function],
      },
    ],
    "type": "section",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Filter
            onChange={[Function]}
          />,
          <ForwardRef(Bootstrap(Row))
            className="justify-content-center"
          >
            <Spinner />
          </ForwardRef(Bootstrap(Row))>,
          false,
          <CommonModal
            onHide={[Function]}
            show={false}
            title="Oops!"
          >
            <strong>
              
            </strong>
          </CommonModal>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "onChange": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": <Spinner />,
            "className": "justify-content-center",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "defaultProps": undefined,
            "propTypes": undefined,
            "render": [Function],
          },
        },
        false,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": <strong>
              
            </strong>,
            "onHide": [Function],
            "show": false,
            "title": "Oops!",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "",
            },
            "ref": null,
            "rendered": "",
            "type": "strong",
          },
          "type": [Function],
        },
      ],
      "type": "section",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
